In [1]:
import numpy as np 
import pandas as pd
import csv

In [2]:
data_path = 'HR_comma_sep.csv'

In [3]:
df = pd.read_csv(data_path)
print(df)

       satisfaction_level  last_evaluation  number_project  \
0                    0.38             0.53               2   
1                    0.80             0.86               5   
2                    0.11             0.88               7   
3                    0.72             0.87               5   
4                    0.37             0.52               2   
...                   ...              ...             ...   
14994                0.40             0.57               2   
14995                0.37             0.48               2   
14996                0.37             0.53               2   
14997                0.11             0.96               6   
14998                0.37             0.52               2   

       average_montly_hours  time_spend_company  Work_accident  left  \
0                       157                   3              0     1   
1                       262                   6              0     1   
2                       272            

In [4]:
df.head(10)

satisfaction_level  last_evaluation  number_project  average_montly_hours  \
0                0.38             0.53               2                   157   
1                0.80             0.86               5                   262   
2                0.11             0.88               7                   272   
3                0.72             0.87               5                   223   
4                0.37             0.52               2                   159   
5                0.41             0.50               2                   153   
6                0.10             0.77               6                   247   
7                0.92             0.85               5                   259   
8                0.89             1.00               5                   224   
9                0.42             0.53               2                   142   

   time_spend_company  Work_accident  left  promotion_last_5years Department  \
0                   3              0     1                      0      sales   
1                   6              0     1                      0      sales   
2                   4              0     1                      0      sales   
3                   5              0     1                      0      sales   
4                   3              0     1                      0      sales   
5                   3              0     1                      0      sales   
6                   4              0     1                      0      sales   
7                   5              0     1                      0      sales   
8                   5              0     1                      0      sales   
9                   3              0     1                      0      sales   

   salary  
0     low  
1  medium  
2  medium  
3     low  
4     low  
5     low  
6     low  
7     low  
8     low  
9     low

In [5]:
df.shape

(14999, 10)

In [6]:
df.index

RangeIndex(start=0, stop=14999, step=1)

In [7]:
df.columns

Index(['satisfaction_level', 'last_evaluation', 'number_project',
       'average_montly_hours', 'time_spend_company', 'Work_accident', 'left',
       'promotion_last_5years', 'Department', 'salary'],
      dtype='object')

In [8]:
df.dtypes

satisfaction_level       float64
last_evaluation          float64
number_project             int64
average_montly_hours       int64
time_spend_company         int64
Work_accident              int64
left                       int64
promotion_last_5years      int64
Department                object
salary                    object
dtype: object

In [9]:
df.rename(columns={
    'last_evaluation':'le'
})

satisfaction_level    le  number_project  average_montly_hours  \
0                    0.38  0.53               2                   157   
1                    0.80  0.86               5                   262   
2                    0.11  0.88               7                   272   
3                    0.72  0.87               5                   223   
4                    0.37  0.52               2                   159   
...                   ...   ...             ...                   ...   
14994                0.40  0.57               2                   151   
14995                0.37  0.48               2                   160   
14996                0.37  0.53               2                   143   
14997                0.11  0.96               6                   280   
14998                0.37  0.52               2                   158   

       time_spend_company  Work_accident  left  promotion_last_5years  \
0                       3              0     1                      0   
1                       6              0     1                      0   
2                       4              0     1                      0   
3                       5              0     1                      0   
4                       3              0     1                      0   
...                   ...            ...   ...                    ...   
14994                   3              0     1                      0   
14995                   3              0     1                      0   
14996                   3              0     1                      0   
14997                   4              0     1                      0   
14998                   3              0     1                      0   

      Department  salary  
0          sales     low  
1          sales  medium  
2          sales  medium  
3          sales     low  
4          sales     low  
...          ...     ...  
14994    support     low  
14995    support     low  
14996    support     low  
14997    support     low  
14998    support     low  

[14999 rows x 10 columns]

In [10]:
df.dtypes

satisfaction_level       float64
last_evaluation          float64
number_project             int64
average_montly_hours       int64
time_spend_company         int64
Work_accident              int64
left                       int64
promotion_last_5years      int64
Department                object
salary                    object
dtype: object

In [11]:
df.values

array([[0.38, 0.53, 2, ..., 0, 'sales', 'low'],
       [0.8, 0.86, 5, ..., 0, 'sales', 'medium'],
       [0.11, 0.88, 7, ..., 0, 'sales', 'medium'],
       ...,
       [0.37, 0.53, 2, ..., 0, 'support', 'low'],
       [0.11, 0.96, 6, ..., 0, 'support', 'low'],
       [0.37, 0.52, 2, ..., 0, 'support', 'low']], dtype=object)

In [12]:
df.to_numpy()

array([[0.38, 0.53, 2, ..., 0, 'sales', 'low'],
       [0.8, 0.86, 5, ..., 0, 'sales', 'medium'],
       [0.11, 0.88, 7, ..., 0, 'sales', 'medium'],
       ...,
       [0.37, 0.53, 2, ..., 0, 'support', 'low'],
       [0.11, 0.96, 6, ..., 0, 'support', 'low'],
       [0.37, 0.52, 2, ..., 0, 'support', 'low']], dtype=object)

In [13]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 14999 entries, 0 to 14998
Data columns (total 10 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   satisfaction_level     14999 non-null  float64
 1   last_evaluation        14999 non-null  float64
 2   number_project         14999 non-null  int64  
 3   average_montly_hours   14999 non-null  int64  
 4   time_spend_company     14999 non-null  int64  
 5   Work_accident          14999 non-null  int64  
 6   left                   14999 non-null  int64  
 7   promotion_last_5years  14999 non-null  int64  
 8   Department             14999 non-null  object 
 9   salary                 14999 non-null  object 
dtypes: float64(2), int64(6), object(2)
memory usage: 1.1+ MB


-DataFrame occupies 1.1+MB memory

-Columns are of types: float64, int64, object

-Total rows= 14999
           
-No any columns contains null values

In [14]:
df.describe()

satisfaction_level  last_evaluation  number_project  \
count        14999.000000     14999.000000    14999.000000   
mean             0.612834         0.716102        3.803054   
std              0.248631         0.171169        1.232592   
min              0.090000         0.360000        2.000000   
25%              0.440000         0.560000        3.000000   
50%              0.640000         0.720000        4.000000   
75%              0.820000         0.870000        5.000000   
max              1.000000         1.000000        7.000000   

       average_montly_hours  time_spend_company  Work_accident          left  \
count          14999.000000        14999.000000   14999.000000  14999.000000   
mean             201.050337            3.498233       0.144610      0.238083   
std               49.943099            1.460136       0.351719      0.425924   
min               96.000000            2.000000       0.000000      0.000000   
25%              156.000000            3.000000       0.000000      0.000000   
50%              200.000000            3.000000       0.000000      0.000000   
75%              245.000000            4.000000       0.000000      0.000000   
max              310.000000           10.000000       1.000000      1.000000   

       promotion_last_5years  
count           14999.000000  
mean                0.021268  
std                 0.144281  
min                 0.000000  
25%                 0.000000  
50%                 0.000000  
75%                 0.000000  
max                 1.000000

We get the summary of each dataframe column easily without the need for explicit computation.

By default it includes only the numerical columns.

Lets fet description of object type only in below cells

In [15]:
df.describe(exclude = ['int', 'float'])

Department salary
count       14999  14999
unique         10      3
top         sales    low
freq         4140   7316

For object type, we get count, unique, top items with frequency.

Data Selction

This selection different techniques to access columns, slice dataframe, get unique column values.

select columns(using[], and dot(.) operator)

iloc

loc

unique()  and

In [16]:
print(df['satisfaction_level'])
print('.................')
print(df.satisfaction_level)

0        0.38
1        0.80
2        0.11
3        0.72
4        0.37
         ... 
14994    0.40
14995    0.37
14996    0.37
14997    0.11
14998    0.37
Name: satisfaction_level, Length: 14999, dtype: float64
.................
0        0.38
1        0.80
2        0.11
3        0.72
4        0.37
         ... 
14994    0.40
14995    0.37
14996    0.37
14997    0.11
14998    0.37
Name: satisfaction_level, Length: 14999, dtype: float64


In [17]:
columns_to_select = ['satisfaction_level', 'left']
df[columns_to_select]

satisfaction_level  left
0                    0.38     1
1                    0.80     1
2                    0.11     1
3                    0.72     1
4                    0.37     1
...                   ...   ...
14994                0.40     1
14995                0.37     1
14996                0.37     1
14997                0.11     1
14998                0.37     1

[14999 rows x 2 columns]

In [18]:
df.iloc[1,1]

0.86

In [19]:
df.iloc[:3,-3:]

promotion_last_5years Department  salary
0                      0      sales     low
1                      0      sales  medium
2                      0      sales  medium

In [20]:
df.loc[0,'satisfaction_level']

0.38

In [21]:
row_level = [0, 1, 2]
column_level = ['promotion_last_5years','Department','salary']
df.loc[row_level, column_level]

promotion_last_5years Department  salary
0                      0      sales     low
1                      0      sales  medium
2                      0      sales  medium

In [22]:
df['left'].unique()

array([1, 0])

In [23]:
df['left'].nunique()

2

In [24]:
df.promotion_last_5years.nunique()

2

In [25]:
df['left'].count()

14999

Aggregations

Methods are:

mean()

std()

sum()

max()

min()

In [26]:
df.describe()

satisfaction_level  last_evaluation  number_project  \
count        14999.000000     14999.000000    14999.000000   
mean             0.612834         0.716102        3.803054   
std              0.248631         0.171169        1.232592   
min              0.090000         0.360000        2.000000   
25%              0.440000         0.560000        3.000000   
50%              0.640000         0.720000        4.000000   
75%              0.820000         0.870000        5.000000   
max              1.000000         1.000000        7.000000   

       average_montly_hours  time_spend_company  Work_accident          left  \
count          14999.000000        14999.000000   14999.000000  14999.000000   
mean             201.050337            3.498233       0.144610      0.238083   
std               49.943099            1.460136       0.351719      0.425924   
min               96.000000            2.000000       0.000000      0.000000   
25%              156.000000            3.000000       0.000000      0.000000   
50%              200.000000            3.000000       0.000000      0.000000   
75%              245.000000            4.000000       0.000000      0.000000   
max              310.000000           10.000000       1.000000      1.000000   

       promotion_last_5years  
count           14999.000000  
mean                0.021268  
std                 0.144281  
min                 0.000000  
25%                 0.000000  
50%                 0.000000  
75%                 0.000000  
max                 1.000000

In [27]:
df.describe(include = ['int', 'float'])

satisfaction_level  last_evaluation  number_project  \
count        14999.000000     14999.000000    14999.000000   
mean             0.612834         0.716102        3.803054   
std              0.248631         0.171169        1.232592   
min              0.090000         0.360000        2.000000   
25%              0.440000         0.560000        3.000000   
50%              0.640000         0.720000        4.000000   
75%              0.820000         0.870000        5.000000   
max              1.000000         1.000000        7.000000   

       average_montly_hours  time_spend_company  Work_accident          left  \
count          14999.000000        14999.000000   14999.000000  14999.000000   
mean             201.050337            3.498233       0.144610      0.238083   
std               49.943099            1.460136       0.351719      0.425924   
min               96.000000            2.000000       0.000000      0.000000   
25%              156.000000            3.000000       0.000000      0.000000   
50%              200.000000            3.000000       0.000000      0.000000   
75%              245.000000            4.000000       0.000000      0.000000   
max              310.000000           10.000000       1.000000      1.000000   

       promotion_last_5years  
count           14999.000000  
mean                0.021268  
std                 0.144281  
min                 0.000000  
25%                 0.000000  
50%                 0.000000  
75%                 0.000000  
max                 1.000000

In [28]:
df.describe(include=['int', 'float']).mean()

satisfaction_level       1875.356433
last_evaluation          1875.424659
number_project           1878.129456
average_montly_hours     2032.124180
time_spend_company       1878.244796
Work_accident            1875.062041
left                     1875.083001
promotion_last_5years    1875.020694
dtype: float64

In [29]:
df['Work_accident'].sum()

2169

In [30]:
df.describe(include=['int', 'float']).std()

satisfaction_level       5302.752785
last_evaluation          5302.725217
number_project           5301.632605
average_montly_hours     5240.043131
time_spend_company       5301.586356
Work_accident            5302.871741
left                     5302.863272
promotion_last_5years    5302.888448
dtype: float64

In [31]:
df['satisfaction_level'].sum()

9191.89

In [32]:
df['satisfaction_level'].max()

1.0

In [33]:
df['satisfaction_level'].min()

0.09

Missing Data:

In this sections we will explore methods available to find missing data and handle missing data in our data:

Methods are:

isnull() or isna()

notnull() or notna()

dropna()

fillna()

isnull() or isna()

Returns True if DataFrame values is null

DataFrame method to get nan values from columns.

It will considered null values if there are nan which is python default marker fro null values.

You can use df.usnull() methods which is simply as alias of df.isna()

In [34]:
df.isnull()

satisfaction_level  last_evaluation  number_project  \
0                   False            False           False   
1                   False            False           False   
2                   False            False           False   
3                   False            False           False   
4                   False            False           False   
...                   ...              ...             ...   
14994               False            False           False   
14995               False            False           False   
14996               False            False           False   
14997               False            False           False   
14998               False            False           False   

       average_montly_hours  time_spend_company  Work_accident   left  \
0                     False               False          False  False   
1                     False               False          False  False   
2                     False               False          False  False   
3                     False               False          False  False   
4                     False               False          False  False   
...                     ...                 ...            ...    ...   
14994                 False               False          False  False   
14995                 False               False          False  False   
14996                 False               False          False  False   
14997                 False               False          False  False   
14998                 False               False          False  False   

       promotion_last_5years  Department  salary  
0                      False       False   False  
1                      False       False   False  
2                      False       False   False  
3                      False       False   False  
4                      False       False   False  
...                      ...         ...     ...  
14994                  False       False   False  
14995                  False       False   False  
14996                  False       False   False  
14997                  False       False   False  
14998                  False       False   False  

[14999 rows x 10 columns]

In [35]:
df.isnull().sum()

satisfaction_level       0
last_evaluation          0
number_project           0
average_montly_hours     0
time_spend_company       0
Work_accident            0
left                     0
promotion_last_5years    0
Department               0
salary                   0
dtype: int64

This means there are no any values 

In [36]:
df.notnull()

satisfaction_level  last_evaluation  number_project  \
0                    True             True            True   
1                    True             True            True   
2                    True             True            True   
3                    True             True            True   
4                    True             True            True   
...                   ...              ...             ...   
14994                True             True            True   
14995                True             True            True   
14996                True             True            True   
14997                True             True            True   
14998                True             True            True   

       average_montly_hours  time_spend_company  Work_accident  left  \
0                      True                True           True  True   
1                      True                True           True  True   
2                      True                True           True  True   
3                      True                True           True  True   
4                      True                True           True  True   
...                     ...                 ...            ...   ...   
14994                  True                True           True  True   
14995                  True                True           True  True   
14996                  True                True           True  True   
14997                  True                True           True  True   
14998                  True                True           True  True   

       promotion_last_5years  Department  salary  
0                       True        True    True  
1                       True        True    True  
2                       True        True    True  
3                       True        True    True  
4                       True        True    True  
...                      ...         ...     ...  
14994                   True        True    True  
14995                   True        True    True  
14996                   True        True    True  
14997                   True        True    True  
14998                   True        True    True  

[14999 rows x 10 columns]

In [37]:
df.notnull().sum()

satisfaction_level       14999
last_evaluation          14999
number_project           14999
average_montly_hours     14999
time_spend_company       14999
Work_accident            14999
left                     14999
promotion_last_5years    14999
Department               14999
salary                   14999
dtype: int64

In [38]:
df.dropna()

satisfaction_level  last_evaluation  number_project  \
0                    0.38             0.53               2   
1                    0.80             0.86               5   
2                    0.11             0.88               7   
3                    0.72             0.87               5   
4                    0.37             0.52               2   
...                   ...              ...             ...   
14994                0.40             0.57               2   
14995                0.37             0.48               2   
14996                0.37             0.53               2   
14997                0.11             0.96               6   
14998                0.37             0.52               2   

       average_montly_hours  time_spend_company  Work_accident  left  \
0                       157                   3              0     1   
1                       262                   6              0     1   
2                       272                   4              0     1   
3                       223                   5              0     1   
4                       159                   3              0     1   
...                     ...                 ...            ...   ...   
14994                   151                   3              0     1   
14995                   160                   3              0     1   
14996                   143                   3              0     1   
14997                   280                   4              0     1   
14998                   158                   3              0     1   

       promotion_last_5years Department  salary  
0                          0      sales     low  
1                          0      sales  medium  
2                          0      sales  medium  
3                          0      sales     low  
4                          0      sales     low  
...                      ...        ...     ...  
14994                      0    support     low  
14995                      0    support     low  
14996                      0    support     low  
14997                      0    support     low  
14998                      0    support     low  

[14999 rows x 10 columns]

In [39]:
df.fillna(5)

satisfaction_level  last_evaluation  number_project  \
0                    0.38             0.53               2   
1                    0.80             0.86               5   
2                    0.11             0.88               7   
3                    0.72             0.87               5   
4                    0.37             0.52               2   
...                   ...              ...             ...   
14994                0.40             0.57               2   
14995                0.37             0.48               2   
14996                0.37             0.53               2   
14997                0.11             0.96               6   
14998                0.37             0.52               2   

       average_montly_hours  time_spend_company  Work_accident  left  \
0                       157                   3              0     1   
1                       262                   6              0     1   
2                       272                   4              0     1   
3                       223                   5              0     1   
4                       159                   3              0     1   
...                     ...                 ...            ...   ...   
14994                   151                   3              0     1   
14995                   160                   3              0     1   
14996                   143                   3              0     1   
14997                   280                   4              0     1   
14998                   158                   3              0     1   

       promotion_last_5years Department  salary  
0                          0      sales     low  
1                          0      sales  medium  
2                          0      sales  medium  
3                          0      sales     low  
4                          0      sales     low  
...                      ...        ...     ...  
14994                      0    support     low  
14995                      0    support     low  
14996                      0    support     low  
14997                      0    support     low  
14998                      0    support     low  

[14999 rows x 10 columns]

Data Manipulation:

Methods are:

-replace()

-astype()

-drop()

sort_values()

groupby()

apply()

In [40]:
df['left'].replace({1:'yes', 0:'no'}, inplace=True)
df['left'].value_counts()

/tmp/ipykernel_6615/2774722963.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['left'].replace({1:'yes', 0:'no'}, inplace=True)


left
no     11428
yes     3571
Name: count, dtype: int64

In [41]:
df_typecast = df.astype({'left':'bool'})
df_typecast.head()

satisfaction_level  last_evaluation  number_project  average_montly_hours  \
0                0.38             0.53               2                   157   
1                0.80             0.86               5                   262   
2                0.11             0.88               7                   272   
3                0.72             0.87               5                   223   
4                0.37             0.52               2                   159   

   time_spend_company  Work_accident  left  promotion_last_5years Department  \
0                   3              0  True                      0      sales   
1                   6              0  True                      0      sales   
2                   4              0  True                      0      sales   
3                   5              0  True                      0      sales   
4                   3              0  True                      0      sales   

   salary  
0     low  
1  medium  
2  medium  
3     low  
4     low

In [42]:
df_drop = df.drop('salary', axis=1)
df_drop.head()

satisfaction_level  last_evaluation  number_project  average_montly_hours  \
0                0.38             0.53               2                   157   
1                0.80             0.86               5                   262   
2                0.11             0.88               7                   272   
3                0.72             0.87               5                   223   
4                0.37             0.52               2                   159   

   time_spend_company  Work_accident left  promotion_last_5years Department  
0                   3              0  yes                      0      sales  
1                   6              0  yes                      0      sales  
2                   4              0  yes                      0      sales  
3                   5              0  yes                      0      sales  
4                   3              0  yes                      0      sales

In [43]:
df_sort = df.sort_values(by='Work_accident', ascending=False)
df_sort.head()
# print(df_sort)

satisfaction_level  last_evaluation  number_project  \
8078                 0.77             0.91               5   
3570                 0.50             0.60               3   
1784                 0.43             0.51               2   
6589                 0.64             0.64               3   
11800                0.67             0.91               2   

       average_montly_hours  time_spend_company  Work_accident left  \
8078                    268                   3              1   no   
3570                    270                   3              1   no   
1784                    149                   3              1  yes   
6589                    234                   3              1   no   
11800                   245                   2              1   no   

       promotion_last_5years  Department  salary  
8078                       0          IT  medium  
3570                       0  management  medium  
1784                       0  accounting  medium  
6589                       0       sales    high  
11800                      1       sales  medium

In [44]:
df_depart = df.groupby('Department').agg({'promotion_last_5years':'sum'})
df_depart

promotion_last_5years
Department                        
IT                               3
RandD                           27
accounting                      14
hr                              15
management                      69
marketing                       43
product_mng                      0
sales                          100
support                         20
technical                       28